File: app/routers/menu.py
from fastapi import APIRouter, Request, Depends, Form
from fastapi.responses import HTMLResponse, RedirectResponse
from sqlalchemy.orm import Session
from typing import Optional
import json

from ..database import get_db
from ..models import MenuCategory, MenuItem

router = APIRouter(prefix="/menu", tags=["menu"])


def require_admin(request: Request):
    if not request.session.get("user_id"):
        return RedirectResponse(url="/login", status_code=302)
    if request.session.get("role") != "admin":
        return RedirectResponse(url="/dashboard", status_code=302)
    return None


def flash(request: Request, message: str, type: str = "success"):
    if "flash" not in request.session:
        request.session["flash"] = []
    request.session["flash"].append({"message": message, "type": type})


# ── GET /menu ──────────────────────────────────────────────────────────────────
@router.get("", response_class=HTMLResponse)
async def menu_index(
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    categories = (
        db.query(MenuCategory)
        .filter(MenuCategory.is_active == True)
        .order_by(MenuCategory.sort_order, MenuCategory.name)
        .all()
    )
    items = (
        db.query(MenuItem)
        .filter(MenuItem.is_active == True)
        .order_by(MenuItem.category_id, MenuItem.sort_order, MenuItem.name)
        .all()
    )

    # Build JSON for Alpine.js
    cat_map = {c.id: c.name for c in categories}
    items_json = json.dumps([
        {
            "id": it.id,
            "name": it.name,
            "desc": it.description or "",
            "price": it.base_price,
            "cat_id": it.category_id,
            "cat_name": cat_map.get(it.category_id, ""),
            "available": it.is_available,
        }
        for it in items
    ], ensure_ascii=False)

    templates = request.app.state.templates
    return templates.TemplateResponse(
        "menu/index.html",
        {
            "request": request,
            "categories": categories,
            "items": items,
            "items_json": items_json,
        },
    )


# ── GET /menu/categories ───────────────────────────────────────────────────────
@router.get("/categories", response_class=HTMLResponse)
async def categories_index(
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    categories = (
        db.query(MenuCategory)
        .filter(MenuCategory.is_active == True)
        .order_by(MenuCategory.sort_order, MenuCategory.name)
        .all()
    )

    templates = request.app.state.templates
    return templates.TemplateResponse(
        "menu/categories.html",
        {"request": request, "categories": categories},
    )


# ── POST /menu/categories ──────────────────────────────────────────────────────
@router.post("/categories")
async def create_category(
    request: Request,
    name: str = Form(...),
    description: str = Form(""),
    sort_order: int = Form(0),
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    cat = MenuCategory(name=name.strip(), description=description.strip() or None, sort_order=sort_order)
    db.add(cat)
    db.commit()
    flash(request, f"Đã thêm danh mục "{name}".")
    return RedirectResponse(url="/menu/categories", status_code=302)


# ── POST /menu/categories/{id}/edit ───────────────────────────────────────────
@router.post("/categories/{cat_id}/edit")
async def edit_category(
    cat_id: int,
    request: Request,
    name: str = Form(...),
    description: str = Form(""),
    sort_order: int = Form(0),
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    cat = db.query(MenuCategory).filter(MenuCategory.id == cat_id).first()
    if cat:
        cat.name = name.strip()
        cat.description = description.strip() or None
        cat.sort_order = sort_order
        db.commit()
        flash(request, f"Đã cập nhật danh mục "{name}".")
    return RedirectResponse(url="/menu/categories", status_code=302)


# ── POST /menu/categories/{id}/delete ─────────────────────────────────────────
@router.post("/categories/{cat_id}/delete")
async def delete_category(
    cat_id: int,
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    cat = db.query(MenuCategory).filter(MenuCategory.id == cat_id).first()
    if cat:
        cat.is_active = False
        db.commit()
        flash(request, f"Đã ẩn danh mục "{cat.name}".")
    return RedirectResponse(url="/menu/categories", status_code=302)


# ── GET /menu/items/new ────────────────────────────────────────────────────────
@router.get("/items/new", response_class=HTMLResponse)
async def new_item_form(
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    categories = (
        db.query(MenuCategory)
        .filter(MenuCategory.is_active == True)
        .order_by(MenuCategory.sort_order, MenuCategory.name)
        .all()
    )

    templates = request.app.state.templates
    return templates.TemplateResponse(
        "menu/form.html",
        {"request": request, "categories": categories, "item": None},
    )


# ── POST /menu/items ───────────────────────────────────────────────────────────
@router.post("/items")
async def create_item(
    request: Request,
    category_id: int = Form(...),
    name: str = Form(...),
    description: str = Form(""),
    base_price: float = Form(...),
    sort_order: int = Form(0),
    is_available: Optional[str] = Form(None),
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    item = MenuItem(
        category_id=category_id,
        name=name.strip(),
        description=description.strip() or None,
        base_price=base_price,
        sort_order=sort_order,
        is_available=is_available == "on",
    )
    db.add(item)
    db.commit()
    flash(request, f"Đã thêm món "{name}".")
    return RedirectResponse(url="/menu", status_code=302)


# ── GET /menu/items/{id}/edit ──────────────────────────────────────────────────
@router.get("/items/{item_id}/edit", response_class=HTMLResponse)
async def edit_item_form(
    item_id: int,
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    item = db.query(MenuItem).filter(MenuItem.id == item_id, MenuItem.is_active == True).first()
    if not item:
        flash(request, "Không tìm thấy món ăn.", "error")
        return RedirectResponse(url="/menu", status_code=302)

    categories = (
        db.query(MenuCategory)
        .filter(MenuCategory.is_active == True)
        .order_by(MenuCategory.sort_order, MenuCategory.name)
        .all()
    )

    templates = request.app.state.templates
    return templates.TemplateResponse(
        "menu/form.html",
        {"request": request, "categories": categories, "item": item},
    )


# ── POST /menu/items/{id}/edit ─────────────────────────────────────────────────
@router.post("/items/{item_id}/edit")
async def update_item(
    item_id: int,
    request: Request,
    category_id: int = Form(...),
    name: str = Form(...),
    description: str = Form(""),
    base_price: float = Form(...),
    sort_order: int = Form(0),
    is_available: Optional[str] = Form(None),
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    item = db.query(MenuItem).filter(MenuItem.id == item_id).first()
    if item:
        item.category_id = category_id
        item.name = name.strip()
        item.description = description.strip() or None
        item.base_price = base_price
        item.sort_order = sort_order
        item.is_available = is_available == "on"
        db.commit()
        flash(request, f"Đã cập nhật món "{name}".")
    return RedirectResponse(url="/menu", status_code=302)


# ── POST /menu/items/{id}/toggle ──────────────────────────────────────────────
@router.post("/items/{item_id}/toggle")
async def toggle_item(
    item_id: int,
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    item = db.query(MenuItem).filter(MenuItem.id == item_id).first()
    if item:
        item.is_available = not item.is_available
        db.commit()
        status = "Có sẵn" if item.is_available else "Tạm hết"
        flash(request, f"Món "{item.name}" → {status}.")
    return RedirectResponse(url="/menu", status_code=302)


# ── POST /menu/items/{id}/delete ──────────────────────────────────────────────
@router.post("/items/{item_id}/delete")
async def delete_item(
    item_id: int,
    request: Request,
    db: Session = Depends(get_db),
):
    redirect = require_admin(request)
    if redirect:
        return redirect

    item = db.query(MenuItem).filter(MenuItem.id == item_id).first()
    if item:
        item.is_active = False
        db.commit()
        flash(request, f"Đã ẩn món "{item.name}".")
    return RedirectResponse(url="/menu", status_code=302)

File: src/routes/menu.js
const express = require('express');
const router = express.Router();
const { q } = require('../db');
const { requireAuth, requireAdmin } = require('../middleware/auth');

// GET / — Danh sách thực đơn
router.get('/', requireAuth, (req, res) => {
  const categories = q.all(
    `SELECT * FROM menu_categories WHERE is_active = 1 ORDER BY sort_order, name`
  );
  const items = q.all(
    `SELECT mi.*, mc.name as category_name
     FROM menu_items mi
     JOIN menu_categories mc ON mc.id = mi.category_id
     WHERE mi.is_active = 1
     ORDER BY mc.sort_order, mi.sort_order, mi.name`
  );
  res.render('menu/index.html', { categories, items });
});

// GET /categories — Quản lý danh mục
router.get('/categories', requireAdmin, (req, res) => {
  const categories = q.all(
    `SELECT mc.*,
       COUNT(CASE WHEN mi.is_active = 1 THEN 1 END) as item_count
     FROM menu_categories mc
     LEFT JOIN menu_items mi ON mi.category_id = mc.id
     WHERE mc.is_active = 1
     GROUP BY mc.id
     ORDER BY mc.sort_order, mc.name`
  );
  res.render('menu/categories.html', { categories });
});

// POST /categories — Thêm danh mục
router.post('/categories', requireAdmin, (req, res) => {
  const { name, sort_order } = req.body;
  if (!name || !name.trim()) {
    res.flash('error', 'Tên danh mục không được để trống');
    return res.redirect('/menu/categories');
  }
  q.run(
    `INSERT INTO menu_categories (name, sort_order) VALUES (?, ?)`,
    name.trim(),
    parseInt(sort_order) || 0
  );
  res.flash('success', `Đã thêm danh mục "${name.trim()}"`);
  res.redirect('/menu/categories');
});

// POST /categories/:id/edit — Sửa danh mục
router.post('/categories/:id/edit', requireAdmin, (req, res) => {
  const { id } = req.params;
  const { name, sort_order } = req.body;
  if (!name || !name.trim()) {
    res.flash('error', 'Tên danh mục không được để trống');
    return res.redirect('/menu/categories');
  }
  q.run(
    `UPDATE menu_categories
     SET name = ?, sort_order = ?, updated_at = datetime('now','localtime')
     WHERE id = ?`,
    name.trim(),
    parseInt(sort_order) || 0,
    id
  );
  res.flash('success', 'Đã cập nhật danh mục');
  res.redirect('/menu/categories');
});

// POST /categories/:id/delete — Xoá danh mục (soft delete)
router.post('/categories/:id/delete', requireAdmin, (req, res) => {
  const { id } = req.params;
  const hasItems = q.get(
    `SELECT 1 FROM menu_items WHERE category_id = ? AND is_active = 1 LIMIT 1`,
    id
  );
  if (hasItems) {
    res.flash('error', 'Danh mục đang có món ăn, không thể xóa');
    return res.redirect('/menu/categories');
  }
  q.run(
    `UPDATE menu_categories SET is_active = 0, updated_at = datetime('now','localtime') WHERE id = ?`,
    id
  );
  res.flash('success', 'Đã xoá danh mục');
  res.redirect('/menu/categories');
});

// GET /items/new — Form thêm món
router.get('/items/new', requireAdmin, (req, res) => {
  const categories = q.all(
    `SELECT * FROM menu_categories WHERE is_active = 1 ORDER BY sort_order, name`
  );
  res.render('menu/form.html', { categories, item: null, action: '/menu/items' });
});

// POST /items — Tạo món mới
router.post('/items', requireAdmin, (req, res) => {
  const { category_id, name, description, base_price, sort_order, is_available } = req.body;
  if (!name || !name.trim()) {
    res.flash('error', 'Tên món không được để trống');
    return res.redirect('/menu/items/new');
  }
  q.run(
    `INSERT INTO menu_items (category_id, name, description, base_price, sort_order, is_available)
     VALUES (?, ?, ?, ?, ?, ?)`,
    parseInt(category_id),
    name.trim(),
    description || '',
    parseFloat(base_price) || 0,
    parseInt(sort_order) || 0,
    is_available ? 1 : 0
  );
  res.flash('success', `Đã thêm món "${name.trim()}"`);
  res.redirect('/menu');
});

// GET /items/:id/edit — Form sửa món
router.get('/items/:id/edit', requireAdmin, (req, res) => {
  const { id } = req.params;
  const item = q.get(`SELECT * FROM menu_items WHERE id = ? AND is_active = 1`, id);
  if (!item) {
    res.flash('error', 'Không tìm thấy món ăn');
    return res.redirect('/menu');
  }
  const categories = q.all(
    `SELECT * FROM menu_categories WHERE is_active = 1 ORDER BY sort_order, name`
  );
  res.render('menu/form.html', {
    item,
    categories,
    action: '/menu/items/' + id + '/edit',
  });
});

// POST /items/:id/edit — Cập nhật món
router.post('/items/:id/edit', requireAdmin, (req, res) => {
  const { id } = req.params;
  const { category_id, name, description, base_price, sort_order, is_available } = req.body;
  if (!name || !name.trim()) {
    res.flash('error', 'Tên món không được để trống');
    return res.redirect('/menu/items/' + id + '/edit');
  }
  q.run(
    `UPDATE menu_items
     SET category_id = ?, name = ?, description = ?, base_price = ?,
         sort_order = ?, is_available = ?, updated_at = datetime('now','localtime')
     WHERE id = ?`,
    parseInt(category_id),
    name.trim(),
    description || '',
    parseFloat(base_price) || 0,
    parseInt(sort_order) || 0,
    is_available ? 1 : 0,
    id
  );
  res.flash('success', 'Đã cập nhật món ăn');
  res.redirect('/menu');
});

// POST /items/:id/toggle — Toggle is_available (JSON, dùng Alpine fetch)
router.post('/items/:id/toggle', requireAuth, (req, res) => {
  const { id } = req.params;
  const item = q.get(`SELECT id, is_available FROM menu_items WHERE id = ? AND is_active = 1`, id);
  if (!item) {
    return res.status(404).json({ success: false, error: 'Không tìm thấy món' });
  }
  const newValue = item.is_available ? 0 : 1;
  q.run(
    `UPDATE menu_items SET is_available = ?, updated_at = datetime('now','localtime') WHERE id = ?`,
    newValue,
    id
  );
  res.json({ success: true, is_available: newValue });
});

// POST /items/:id/delete — Xoá món (soft delete)
router.post('/items/:id/delete', requireAdmin, (req, res) => {
  const { id } = req.params;
  q.run(
    `UPDATE menu_items SET is_active = 0, updated_at = datetime('now','localtime') WHERE id = ?`,
    id
  );
  res.flash('success', 'Đã xoá món ăn');
  res.redirect('/menu');
});

module.exports = router;
